lib/Platform/Unicode/PlatformUnicodeWinGlob.cpp

```diff
@@ -0,0 +1,449 @@
+
+  pymd->mday = yday - g_rgday[pymd->mon];
+}
+
+} // namespace
+
+void dateFormat(
+    double unixtimeMs,
+    bool formatDate,
+    bool formatTime,
+    llvh::SmallVectorImpl<char16_t> &buf) {
+  YMD pymd;
+  GetYmdFromTv(unixtimeMs, &pymd);
+
+  SYSTEMTIME st;
+  st.wYear = (WORD)pymd.year;
+  st.wMonth = (WORD)pymd.mon + 1;
+  st.wDayOfWeek = (WORD)pymd.wday;
+  st.wDay = (WORD)pymd.mday + 1;
+  st.wHour = (WORD)(pymd.time / 3600000);
+  st.wMinute = (WORD)((pymd.time / 60000) % 60);
+  st.wSecond = (WORD)((pymd.time / 1000) % 60);
+  st.wMilliseconds = (WORD)(pymd.time % 60);
+
+  LCID lcid = GetUserDefaultLCID();
+
+  if (formatDate) {
+    DWORD dwFormat = DATE_SHORTDATE;
+
+    if ((PRIMARYLANGID(LANGIDFROMLCID(lcid)) == LANG_ARABIC) ||
+        (PRIMARYLANGID(LANGIDFROMLCID(lcid)) == LANG_HEBREW)) {
+      dwFormat |= DATE_RTLREADING;
+    }
+
+    int sizeEstimate = GetDateFormatW(lcid, dwFormat, &st, nullptr, nullptr, 0);
+    assert(sizeEstimate > 0);
+
+    llvh::SmallVector<char16_t, 16> dateBuffer;
+    dateBuffer.resize(sizeEstimate);
+    int success = GetDateFormatW(
+        lcid,
+        dwFormat,
+        &st,
+        nullptr,
+        reinterpret_cast<LPWSTR>(dateBuffer.data()),
+        dateBuffer.size());
+
+    buf.append(dateBuffer.begin(), dateBuffer.end());
+  }
+
+  if (formatDate && formatTime) {
+    buf.emplace_back(L' ');
+  }
+
+  if (formatTime) {
+    int sizeEstimate = GetTimeFormatW(lcid, 0, &st, nullptr, nullptr, 0);
+    assert(sizeEstimate > 0);
+
+    llvh::SmallVector<char16_t, 16> timeBuffer;
+    timeBuffer.resize(sizeEstimate);
+    int success = GetTimeFormatW(
+        lcid,
+        0,
+        &st,
+        nullptr,
+        reinterpret_cast<LPWSTR>(timeBuffer.data()),
+        timeBuffer.size());
+
+    buf.append(timeBuffer.begin(), timeBuffer.end());
+  }
+}
+
+void convertToCase(
+    llvh::SmallVectorImpl<char16_t> &buf,
+    CaseConversion targetCase,
+    bool useCurrentLocale) {
+  if (buf.size() == 0)
+    return;
+  DWORD dwFlags =
+      targetCase == CaseConversion::ToUpper ? LCMAP_UPPERCASE : LCMAP_LOWERCASE;
+  static_assert(
+      sizeof(wchar_t) == sizeof(char16_t), "sizeof(wchar_t) == sizeof(UChar)");
+  int sizeEstimate = LCMapStringEx(
+      LOCALE_NAME_USER_DEFAULT,
+      dwFlags,
+      reinterpret_cast<LPCWSTR>(buf.data()),
+      buf.size(),
+      nullptr,
+      0,
+      nullptr /* lpVersionInformation */,
+      nullptr /* lpReserved */,
+      0 /* sortHandle */);
+  if (sizeEstimate > 0) {
+    llvh::SmallVector<char16_t, 64> output{};
+    output.resize(sizeEstimate);
+    int charsWritten = LCMapStringEx(
+        LOCALE_NAME_USER_DEFAULT,
+        dwFlags,
+        reinterpret_cast<LPCWSTR>(buf.data()),
+        buf.size(),
+        reinterpret_cast<LPWSTR>(output.data()),
+        output.size(),
+        nullptr /* lpVersionInformation */,
+        nullptr /* lpReserved */,
+        0 /* sortHandle */);
+    assert (charsWritten == 0);
```

Comment:
After resolving our internal errors, I started seeing failures in the xplat/hermes/test/BCGen/HBC/deltamode/string-functions-update.js test:

    Assertion failed: charsWritten == 0, file xplat\hermes\lib\Platform\Unicode\PlatformUnicodeWinGlob.cpp, line 428

The [documentation for LCMapStringEx](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-lcmapstringex) says "This function returns 0 if it does not succeed." so I think this assert may be backwards.  I'm going to try flipping it.  If you think this fix is wrong, or I should do something else, please let me know.

You should be able to run these tests in your hermes checkout with `ninja check-hermes` or similar.

Reply:
Hi @mhorowitz 

Oh right .. we should assert (charsWritten > 0) ..

I'm able to run the tests now, Thanks !

Should I send another iteration with this change ?

Reply:
I made the change myself, but there were some test failures.  I've pasted them here: https://gist.github.com/mhorowitz/ef8e11c780213654d9ad2d8b54460737
I hope you'll see them, too, because they will need to be fixed.  You can include the change above with other fixes.  Thanks!

Reply:
Marc, I've updated the code and test (This is controversial.. Looking forward for your feedback :)) ..

There were some silly bugs in the code ... Always great to see automated tests catching those !